neps/nep-0539.md

```diff
@@ -0,0 +1,741 @@
+
+    - Problem 1: This requires `N` gas numbers in each chunk header, or `N*N`
+      numbers per block, where `N` is the number of shards.
+    - Problem 2: We did not have time to simulate it properly. But on paper, it
+      seems each individual delayed queue can still grow indefinitely as the
+      number of shards in the system grows.
+
+5. Smartly track and limit buffer space across different shards. Only accept new
+   transactions if enough buffer space can be reserved ahead of time.
+
+    - Problem 1: Without knowing which shards a transaction touches and how
+      large receipts will be, we have to pessimistically reserve more space than
+      most receipts will actually need.
+    - Problem 2: If buffer space is shared globally, individual queues can still
+      grow really large, even indefinitely if we assume the number of shards
+      grows over time.
+    - Problem 3: If buffer space is on a per-shard basis, we run into deadlocks
+      when two shards have no more space left but both need to send to the other
+      shard to make progress.
+
+6. Require users to define which shards their transactions will touch and how
+   much gas is burnt in each. Then use this information for global scheduling
+   such that congestion is impossible.
+
+    - Problem 1: This requires lots of changes across the infrastructure stack.
+      It would take too long to implement as we are already facing congestion
+      problems today.
+    - Problem 2: This would have a strong impact on usability and it is unclear
+      if gas usage estimating services could close the gap to make it
+      acceptable.
+
+
+
+## Future possibilities
+
+While this proposal treats all requests the same, it sets the base for a proper
+transaction priority implementation. We co-designed this proposal with
+[NEP-541](https://github.com/near/NEPs/pull/541), which adds a transaction
+priority fee. On a very high level, the fee is used to auction off a part of the
+available gas per chunk to the highest bidders.
+
+We also expect that this proposal alone will not be the final solution for
+congestion control. Rather, we just want to build a solid foundation in this NEP
+and allow future optimization to take place on top of it.
+
+For example, estimations on how much gas is burnt on each shard could help with
+better load distribution in some cases.
+
+We also forsee that the round-robin scheduling of shard allowed to forward even
+under full congestion is not perfect. It is a key feature to make deadlocks
+provably impossible, since every shard is guaranteed to make a minimum progress
+after N rounds. But it could be beneficial to allocate more bandwidth to shards
+that actually have something to forward, or perhaps it would be better to stop
+forwarding anything for a while. The current proposal allows chunk producers to
```

Comment:
> The current proposal allows chunk producers to experiment with this without a protocol version change.

True, but practically speaking, chunk producers will use the default implementation in nearcore. So we need to make sure that the signals that we provide (e.g. `allowed_shard`) are enough to iterate to a reasonable policy that deals with the congestion pattern seen on mainnet today.

In my understanding, round-robin on average introduces N/2 = 3 blocks of latency (where N=6 is the number of shards) to cross-contract calls. This could be noticeable for some large contracts like `sweat` and `sweat.claim` that have a lot of communication and are located on different shards that are often congested.

I wonder if there is some way to make the behavior of `allowed_shard` smoother (e.g. allowing multiple senders in each round).

Reply:
Actually, maybe under congestion, the 3 blocks delay is not that big of a deal on top of 20 blocks delay from the delayed receipts queue. Do simulations that you ran provide any insights here?

Reply:
Yes, we do think round-robin is the best choice in the short-term. We just also want to give the protocol team the ability to experiment with it and change it, once we see how congestion control manifests itself on mainnet.

There are main ideas why we think round-robin is best.

1. This is a last-resort measure. The linear reduction of forwarding limits is designed to prevent it from happening in the first place. If we hit it, we are badly congested.
2. If we hit this round-robin because of incoming congestion, we don't introduce extra latency. Anything forwarded is waiting behind a long delayed queue anyway, it might as well wait in outgoing buffers.
3. If we hit it because of outgoing or memory congestion, we are in trouble and are much more concerned about deadlock prevention than about minimizing latencies. The only alternative solution we know of that does not deadlock is a global transaction stop. But global-transaction stop leads to terrible fairness problems, where a single heavily used shard affects all shards and can even cause them to become congested as well due to reduced capacity on them.
4. Allowing multiple shards to send at once quickly leads to conflicts with the 300 Tgas max gas value of a receipt. We went with the simplest possible solution: a single allowed shard.

Reply:
> Actually, maybe under congestion, the 3 blocks delay is not that big of a deal on top of 20 blocks delay from the delayed receipts queue. Do simulations that you ran provide any insights here?

As far as simulations go, it's kind of hard to get into this phase and stay in it. Once in round-robin, it often goes in-and-out of it every few blocks.

"Relayed hot" in the spreadsheet is probably the best example, which has some extra graphs at the bottom of the document. Using the limits we have in the proposal, the median transaction delay is at 19 blocks, the 90th percentile at 24 blocks. This includes 2 cross-shard hops but I'm pretty sure only one is operating under round-robin from time to time. So yeah, 3 blocks isn't a huge deal.